runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/commits/StreamingEngineWorkCommitter.java

```diff
@@ -0,0 +1,209 @@
+        commitWorkStreamFactory, numCommitSenders, onCommitComplete);
+  }
+
+  @Override
+  @SuppressWarnings("FutureReturnValueIgnored")
+  public void start() {
+    if (!commitSenders.isShutdown()) {
+      for (int i = 0; i < numCommitSenders; i++) {
+        commitSenders.submit(this::streamingCommitLoop);
+      }
+    }
+  }
+
+  @Override
+  public void commit(Commit commit) {
+    commitQueue.put(commit);
+  }
+
+  @Override
+  public long currentActiveCommitBytes() {
+    return activeCommitBytes.get();
+  }
+
+  @Override
+  public void stop() {
+    if (!commitSenders.isTerminated() || !commitSenders.isShutdown()) {
+      commitSenders.shutdown();
+      try {
+        commitSenders.awaitTermination(10, TimeUnit.SECONDS);
+      } catch (InterruptedException e) {
+        LOG.warn("Could not shut down commitSenders gracefully, forcing shutdown.", e);
+      }
+      commitSenders.shutdownNow();
+    }
+  }
+
+  @Override
+  public int parallelism() {
+    return numCommitSenders;
+  }
+
+  private void streamingCommitLoop() {
+    @Nullable Commit initialCommit = null;
+    while (true) {
+      if (initialCommit == null) {
+        try {
+          // Block until we have a commit or are shutting down.
+          initialCommit = commitQueue.take();
+        } catch (InterruptedException e) {
+          continue;
+        }
+      }
+
+      if (initialCommit.work().isFailed()) {
+        onCommitComplete.accept(CompleteCommit.forFailedWork(initialCommit));
+        initialCommit = null;
+        continue;
+      }
+
+      try (CloseableStream<CommitWorkStream> closeableCommitStream =
+          commitWorkStreamFactory.get()) {
+        CommitWorkStream commitStream = closeableCommitStream.stream();
+        if (!tryAddToCommitStream(initialCommit, commitStream)) {
+          throw new AssertionError("Initial commit on flushed stream should always be accepted.");
+        }
+        // Batch additional commits to the stream and possibly make an un-batched commit the next
+        // initial commit.
+        initialCommit = batchCommitsToStream(commitStream);
+        commitStream.flush();
+      } catch (Exception e) {
+        LOG.error("Error occurred fetching a CommitWorkStream.", e);
+      }
+    }
+  }
+
+  /** Adds the commit to the commitStream if it fits, returning true if it is consumed. */
+  private boolean tryAddToCommitStream(Commit commit, CommitWorkStream commitStream) {
+    Preconditions.checkNotNull(commit);
+    commit.work().setState(Work.State.COMMITTING);
+    activeCommitBytes.addAndGet(commit.getSize());
+    boolean isCommitSuccessful =
+        commitStream.commitWorkItem(
+            commit.computationId(),
+            commit.request(),
+            (commitStatus) -> onCommitComplete.accept(CompleteCommit.create(commit, commitStatus)));
+
+    if (!isCommitSuccessful) {
+      commit.work().setState(Work.State.COMMIT_QUEUED);
+    }
+
+    activeCommitBytes.addAndGet(-commit.getSize());
```

Comment:
activeCommitBytes is no longer tracking the bytes until they complete, just until they are added on the stream

You should:
- move the decrement into the commitWorkItem ondonefn above.  That handles the case it was added successfully, tracking bytes until the commit response is recieved
- decrement here only if !isCommitSuccesful to cover the case we're keeping it until next time.

Reply:
done